src/Workspaces/Core/Portable/Storage/CloudCache/ProjectCacheContainerKey.cs

```diff
@@ -0,0 +1,97 @@
+        /// container keys when called repeatedly for the same documents.
+        /// </summary>
+        private readonly ConditionalWeakTable<TextDocumentState, StrongBox<CloudCacheContainerKey?>> _documentToContainerKey = new();
+        private readonly ConditionalWeakTable<TextDocumentState, StrongBox<CloudCacheContainerKey?>>.CreateValueCallback _documentToContainerKeyCallback;
+
+        public ProjectCacheContainerKey(string relativePathBase, ProjectKey projectKey)
+        {
+            ContainerKey = CreateProjectContainerKey(relativePathBase, projectKey);
+
+            _documentToContainerKeyCallback = ds => new(CreateDocumentContainerKey(
+                relativePathBase, DocumentKey.ToDocumentKey(projectKey, ds)));
+        }
+
+        public CloudCacheContainerKey? GetValue(TextDocumentState state)
+            => _documentToContainerKey.GetValue(state, _documentToContainerKeyCallback).Value;
+
+        public static CloudCacheContainerKey? CreateProjectContainerKey(
+            string relativePathBase, ProjectKey projectKey)
+        {
+            // Creates a container key for this project.  THe container key is a mix of the project's name, relative
+            // file path (to the solution), and optional parse options.
+
+            // If we don't have a valid solution path, we can't store anything.
+            if (string.IsNullOrEmpty(relativePathBase))
+                return null;
+
+            // We have to have a file path for this project
+            if (string.IsNullOrEmpty(projectKey.FilePath))
+                return null;
+
+            // The file path has to be relative to the solution path.
+            var relativePath = PathUtilities.GetRelativePath(relativePathBase, projectKey.FilePath!);
+            if (relativePath == projectKey.FilePath)
+                return null;
+
+            var dimensions = EmptyDimensions
+                .Add($"{nameof(ProjectKey)}.{nameof(ProjectKey.Name)}", projectKey.Name)
+                .Add($"{nameof(ProjectKey)}.{nameof(ProjectKey.FilePath)}", relativePath)
+                .Add($"{nameof(ProjectKey)}.{nameof(ProjectKey.ParseOptionsChecksum)}", projectKey.ParseOptionsChecksum.ToString());
+
+            return new CloudCacheContainerKey("Roslyn.Project", dimensions);
+        }
+
+        public static CloudCacheContainerKey? CreateDocumentContainerKey(
+            string relativePathBase,
+            DocumentKey documentKey)
+        {
+            // See if we can get a project key for this info.  If not, we def can't get a doc key.
+            var projectContinerKey = CreateProjectContainerKey(relativePathBase, documentKey.Project);
+            if (projectContinerKey == null)
+                return null;
+
+            // We have to have a file path for this document
+            if (string.IsNullOrEmpty(documentKey.FilePath))
+                return null;
+
+            // The file path has to be relative to the solution path.
```

Comment:
"the solution path"

Our recommendation is that your paths be relative to the base path we give you from `ICacheService`, which may be the solution dir or the root of the git repo.

Reply:
it is.  that's just a faulty comment.  will update.